src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            var description = await service.GetProvider(roslynItem).GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(EditorCompletion.CompletionTrigger trigger, SnapshotPoint triggerLocation)
+        {
+            switch (trigger.Reason)
+            {
+                case EditorCompletion.CompletionTriggerReason.Invoke:
+                case EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique:
+                    return RoslynTrigger.Invoke;
+                case EditorCompletion.CompletionTriggerReason.Insertion:
+                    return RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Deletion:
+                    return RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Backspace:
+                    var snapshotBeforeEdit = trigger.ViewSnapshotBeforeTrigger;
+                    char characterRemoved;
+                    if (triggerLocation.Position >= 0 && triggerLocation.Position < snapshotBeforeEdit.Length)
+                    {
+                        // If multiple characters were removed (selection), this finds the first character. 
+                        // Maybe it should be re-considered to find the last removed character.
+                        characterRemoved = snapshotBeforeEdit[triggerLocation.Position];
+                    }
+                    else
+                    {
+                        characterRemoved = (char)0;
+                    }
+
+                    return RoslynTrigger.CreateDeletionTrigger(characterRemoved);
+                case EditorCompletion.CompletionTriggerReason.SnippetsMode:
+                    return new RoslynTrigger(CompletionTriggerKind.Snippets);
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(trigger.Reason);
+        }
+
+        private EditorCompletion.CompletionItem Convert(
+            Document document,
+            RoslynCompletionItem roslynItem,
+            CompletionService completionService,
+            Dictionary<string, EditorCompletion.CompletionFilter> filterCache)
+        {
+            var imageId = roslynItem.Tags.GetFirstGlyph().GetImageId();
+            var filters = GetFilters(roslynItem, filterCache);
+
+            if (!roslynItem.Properties.TryGetValue(InsertionText, out var insertionText))
+            {
+                insertionText = roslynItem.DisplayText;
```

Comment:
what is going on here?  This is making the commit manager know internal implementation details about how specific providers/items encode things...

Reply:
roslynItem generated by providers can contain an insertionText in a property bag.
Try using it first.

Added a comment

---
In reply to: [222083519](https://github.com/dotnet/roslyn/pull/29016#discussion_r222083519) [](ancestors = 222083519)

Reply:
but they can also use any other mechanism they want to store their insertion text.  they can even compute it dynamically.  how do we account for that?

Reply:
We create roslynItem by calling providers. Providers might create the insertion text already and put it into the property bag. Here is what they do: [http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/Providers/SymbolCompletionItem.cs,34](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/Providers/SymbolCompletionItem.cs,34)

I think we should not calculate the insertion text once again if it is already calculated.

---
In reply to: [222491666](https://github.com/dotnet/roslyn/pull/29016#discussion_r222491666) [](ancestors = 222491666)

Reply:
Sorry, my point is not coming across:

You're depending on an implementation detail of certain providers.  Those providers could change how they compute the insertion text at any point in the future.  They might totally compute it dynamically.  Or they may stick it into a different key.  You can't expect or depend on that being there.  Furthermore, any other provider which uses this key may want it for a different purpose.  

The property bag here is a private impl detail.  You shouldn't be looking into it here.

Reply:
Thank you! Then, I could not find that InsertionText is public. There are some protected methods like GetDisplayAndInsertionText in some providers. Do you mean we should make them public and use them?

---
In reply to: [222516610](https://github.com/dotnet/roslyn/pull/29016#discussion_r222516610) [](ancestors = 222516610)

Reply:
No.  These are again implementation details :)  you're at a level that has no concept of how the individual providers are implemented (or even that there are providers at all!).  You need to defer to the CompletionService if you want to get this information.  The only way to know what text will actually be inserted is to call CompletionService.GetChangeAsync and see what actual text the service says will be inserted for that item along with that specific typed char.  

Note: this information cannot be computed when the completion item is created.  The insertion text *depends* on which character was used to commit the item.  Many providers/items compute this on-demand.

Reply:
Makes sense. Thank you! We will not use the insertionText. If we have to commit, we will run GetChangesAsync of provider and get the InsertionText from it. 

Let us consider this InsertionText as a best attempt to provide it to editor in case if somebody else will need it.

---
In reply to: [222518231](https://github.com/dotnet/roslyn/pull/29016#discussion_r222518231) [](ancestors = 222518231)

Reply:
Sounds good.  @AmadeusW What do you guys use insertion text for?  Given that the language is what is deferred to when actually commiting, how does it come into play?  Thanks!